sphinx/builders/html.py

```diff
@@ -404,15 +404,22 @@ def get_doc_context(self, docname, body, metatags):
         # title rendered as HTML
         title = self.env.longtitles.get(docname)
         title = title and self.render_partial(title)['title'] or ''
+
+        # Suffix for the document
+        source_suffix = path.splitext(self.env.doc2path(docname))[1]
+
         # the name for the copied source
-        sourcename = self.config.html_copy_source and docname + '.txt' or ''
+        if self.config.html_copy_source:
+            if self.config.html_sourcelink_keep_suffix:
+                sourcename = docname + source_suffix
+            else:
+                sourcename = docname + '.txt'
```

Comment:
Sure, that would be possible, but I guess since this would be a somewhat breaking change, this should be decided by someone else, e.g. @birkenfeld @shimizukawa @tk0miya ...?

I think `html_sourcelink_add_txt` would be enough.

Note, however, that this will look strange if the source suffix was already `.txt`. This would lead to something like `yetanotherinput.txt.txt`. Is this desirable?
Or should there be an exception to not add `.txt` if it's already there?